src/aspire/orientation/commonline_sync.py

```diff
@@ -1,31 +1,411 @@
 import logging
+import numpy as np
 
-from aspire.orientation import Orient3D
+from scipy.spatial.transform import Rotation
+import scipy.sparse as sparse
+
+from aspire.orientation import CLOrient3D
+from aspire.utils import ensure
 
 logger = logging.getLogger(__name__)
 
 
-class CommLineSync(Orient3D):
+class CommLineSync(CLOrient3D):
     """
-    Define a class to estimate 3D orientations using Synchronization described as below:
-    Y. Shkolnisky, and A. Singer, Viewing Direction Estimation in Cryo-EM Using Synchronization,
+    Define a class to estimate 3D orientations using synchronization matrix and voting method.
+
+    The related publications are listed as below:
+    Y. Shkolnisky, and A. Singer,
+    Viewing Direction Estimation in Cryo-EM Using Synchronization,
     SIAM J. Imaging Sciences, 5, 1088-1110 (2012).
 
+    A. Singer, R. R. Coifman, F. J. Sigworth, D. W. Chester, Y. Shkolnisky,
+    Detecting Consistent Common Lines in Cryo-EM by Voting,
+    Journal of Structural Biology, 169, 312-322 (2010).
     """
-    def __init__(self, src):
+    def __init__(self, src, n_rad=None, n_theta=None):
         """
-        constructor of an object for estimating 3D orientations
+        Initialize an object for estimating 3D orientations using synchronization matrix
+
+        :param src: The source object of 2D denoised or class-averaged images with metadata
+        :param n_rad: The number of points in the radial direction
+        :param n_theta: The number of points in the theta direction
         """
-        pass
+        super().__init__(src, n_rad=n_rad, n_theta=n_theta)
+        self.syncmatrix = None
 
-    def estimate(self):
+    def estimate_rotations(self):
         """
-        perform estimation of orientations
+        Estimate orientation matrices for all 2D images using synchronization matrix
         """
-        pass
+        if self.syncmatrix is None:
+            self.syncmatrix_vote()
+
+        S = self.syncmatrix
+        sz = S.shape
+        ensure(sz[0] == sz[1], 'syncmatrix must be a square matrix.')
+        ensure(sz[0] % 2 == 0, 'syncmatrix must be a square matrix of size 2Kx2K.')
+
+        n_img = sz[0] // 2
+
+        # S is a 2Kx2K matrix (K=n_check), containing KxK blocks of size 2x2.
+        # The [i,j] block is given by [r11 r12; r12 r22], where
```

Comment:
Yes. It can be arranged in the way you mentioned. But the current shape might be more directly to pass it the function for eigenvalues and genvectors such as `sparse.linalg.eigs(S, 10)`. I will keep it unless we have other  better reasons.

Reply:
The better reason is about 100 lines below here.

Reply:
We should be rewriting this section anyway per the notes I left. To change that single line is trivial.

Reply:
The 2K x 2K presentation is more consistent with the paper. If we do the change to KxKx2x2, then we need to rearrange it for the sparse.linalg.eigs function. I think this is less critical in the PR.

Reply:
I believe you would do this `eigs( swapaxes(S, 1,2).reshape(2*k, 2*k) )`. Then when using you get this tasty syntax:

`S[i,j] = thing` instead of a mess; but I could be wrong. I think that swap requires the same order of computational steps as computing the indices below (that was a joke, well sort of, sometimes jokes are true). Joakim can decide I guess. Usually we'd right a util function to compute indices that look like that, fwiw.

Reply:
we discussed making S shape as k,2,k,2 then we can pass S.reshape(2*k, 2*k)

Reply:
Done in commit 7d3af4d.